flang/lib/Lower/Bridge.cpp

```diff
@@ -340,21 +340,22 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         llvm::None);
   }
 
-  bool createHostAssociateVarClone(
-      const Fortran::semantics::Symbol &sym) override final {
+  bool createHostAssociateVarClone(const Fortran::semantics::Symbol &sym,
+                                   bool init) override final {
     auto loc = genLocation(sym.name());
+    mlir::Type symType = genType(sym);
     const auto *details = sym.detailsIf<Fortran::semantics::HostAssocDetails>();
     assert(details != nullptr && "No host-association found");
     const Fortran::semantics::Symbol &hsym = details->symbol();
     Fortran::lower::SymbolBox hsb = lookupSymbol(hsym);
 
     auto allocate = [&](llvm::ArrayRef<mlir::Value> shape,
                         llvm::ArrayRef<mlir::Value> typeParams) -> mlir::Value {
-      mlir::Type symType = genType(sym);
-      return builder->allocateLocal(
+      auto allocVal = builder->allocateLocal(
           loc, symType, mangleName(sym), toStringRef(sym.GetUltimate().name()),
           shape, typeParams,
           sym.GetUltimate().attrs().test(Fortran::semantics::Attr::TARGET));
```

Comment:
I think for a first approximation (and so we can identify them), we should add an attribute to the `fir.alloca` instruction to denote that it is pinned in terms of control flow (if we aren't already doing so). That should be sufficient until we nail down the stack frame management semantics in FIR.

Reply:
I have added a unit attribute `pinned` to the AllocaOp. Builders were updated to always specify the attribute and hence required adding the attribute at a few different places. Also added the attribute in 50+ tests. Although mechanical this makes the patch big.

Besides OpenMP privatisation, the createTemporary function is also updated to use the `pinned` attribute since that function hoists allocas to the entry block.
https://github.com/flang-compiler/f18-llvm-project/blob/ba3942fab13b47423d517efc273f89556e5fe8e8/flang/lib/Optimizer/Builder/FIRBuilder.cpp#L194

Reply:
The addition of the `pinned` attribute was moved to a separate PR and approved and merged. https://github.com/flang-compiler/f18-llvm-project/pull/1065